File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
       Map<Integer, Integer> map = new HashMap<>();
       //map: key -> element in the array , value -> the frequency of that element
       for(int i=0; i<nums.length; i++){
        map.put(nums[i], map.getOrDefault(nums[i], 0)+1);
       } 

        //buckets: index -> the frequency , value -> a list of all elements having that frequency
       List<Integer>[] buckets = new ArrayList[nums.length+1];
       for(int x : map.keySet()){
        if(buckets[map.get(x)]==null){
            buckets[map.get(x)] = new ArrayList<>();
        }
        buckets[map.get(x)].add(x);
       }

        int count = k;
        int[] ans = new int[k];
        int idx=0;
       for(int i=nums.length; i>=1; i--){
            if(buckets[i]==null) continue;
            for(int j=0; j<buckets[i].size(); j++){
                count--;
                if(count<0) return ans;;
                ans[idx] = buckets[i].get(j);
                idx++;
            }
       }
       
       return ans;
    }
}

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        map = dict()
        for num in nums:
            map[num] = map.get(num, 0) + 1

        # [1:3; 2:2; 3:1]
        # key: the element itself
        # value: frequency of that element

        buckets = [[] for _ in range(len(nums)+1)]  
        # index will be the frequency, value will be the element

        res = []
        for key, value in map.items():
            buckets[value].append(key)
        
        # [[], [3], [2], [1], [], [], []]
        for i in range(len(buckets) - 1, -1, -1):
            for num in buckets[i]:
                res.append(num)
            if len(res) == k:
                return res



        